keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,430 @@
+
+Allowing Resources to be changed without recreating the Pod or restarting the
+Containers addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource resize is not possible.
+* Secondary: allow users to specify which Pods and Containers can be resized
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed in-place resource resizing. This should be handled by actors
+which initiated the resizing.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart,
+* updating extended resources or any other resource types besides CPU, memory.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to Container resources requests and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodSpec is extended with information about actually allocated Container resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration, denoting
+  **desired** state of the Pod resources,
+* PodSpec.Container.ResourceBindings (new object) denotes **actual** resources
+  allocated to the Pod by Scheduler,
+* PodStatus.ContainerStatus.ResourcesAllocated (new object) shows the resources
+  held by the Pod and its Containers.
+
+In order to determine the state of a Pod resource update, we add two new
+PodConditions named PodResourcesBound and PodResizeSuccess.
+
+Scheduler sets PodResourcesBound PodCondition, and it can have the following values:
+* Status: false  - Resize request has not been allocated by Scheduler,
+  - Reason: AwaitingPreemption - Lower priority Pods are being pre-empted,
+  - Reason: FailedNodeCapacity - Scheduler determined Node does not have room,
+* Status: true - Scheduler has completed allocating the resize request,
+  - Reason: Accepted - Pod resize approved by Scheduler.
+
+PodResizeSuccess PodCondition can have the following values:
+* Status: true  - Last resize request was successful,
+  - Reason: Success - Pod and its Containers were resized successfully,
+* Status: false - Last resize request failed,
+  - Reason: FailedNodeCapacity - Kubelet determined Node does not have room,
+  - Reason: RejectedResourceQuota - Resize request exceeds ResourceQuota,
+  - Reason: RejectedQoSClassChange - Resize request changes QoS class,
+  - Reason: Retrying - Controller initiated a retry.
+
+#### Container Restart Policy
+
+To provide some fine-grained control to the user,
+PodSpec.Container.ResourceRequirements is extended with ResizePolicy flag
+for each resource type (CPU, memory) :
+* NoRestart - the default value; resize the Container without restarting it,
```

Comment:
i am not sure if this is backwards compatible.  are more applications going to require a restart versus not a restart?  wouldn't the default vary by resource type?  for cpu, i could buy the default being NoRestart, but for memory, i think the more globally common default is to restart.

Reply:
i don't have data to answer this question. the idea with this KEP however is to resize without restarts, naturally NoRestart felt right as the default value. i'm fine with it either ways for memory.

Reply:
We are going with common default for both CPU and memory per the [below discussion](https://github.com/kubernetes/enhancements/pull/686#discussion_r297423537). We can revisit if we find that RestartContainer is desirable default for memory.